include/deal.II/base/quadrature_point_data.h

```diff
@@ -0,0 +1,519 @@
+template <typename CELL, typename DATA>
+class QPManagerBase : public Subscriptor
+{
+public:
+  /**
+   * Constructor.
+   */
+  QPManagerBase();
+
+  /**
+   * Destructor.
+   */
+  virtual ~QPManagerBase();
+
+  /**
+   * Initialize quadrature point data for @p cell with @p n_q_points .
+   */
+  void initialize(const CELL &cell,
+                  const unsigned int n_q_points);
+
+  /**
+   * Get reference to the data located at @p cell .
+   */
+  std::vector<DATA> &get_data(const CELL &cell);
+
+  /**
+   * Get constant reference to the data located at @p cell .
+   */
+  const std::vector<DATA> &get_data(const CELL &cell) const;
+
+  /**
+   * Unpack the data stored in @p values_at_qp of @p cell into the underlying
+   * DATA object at each quadrature point.
+   */
+  void unpack_to_cell_data(const CELL &cell,
+                           const FullMatrix<double> &values_at_qp);
+
+  /**
+   * The opposite of the function above. Pack DATA from each quadrature point of the @p cell into @p matrix_data .
+   */
+  void pack_cell_data(const CELL &cell,
+                      FullMatrix<double> &matrix_data);
+
+protected:
+  /**
+   * A map to store a vector of data at the cell.
+   */
+  std::map<CELL, std::vector<DATA> > map;
+
+};
+
+namespace distributed
+{
+
+  /**
+   * A class for storing and managing quadrature points data in with parallel::distributed::Triangulation .
+   *
+   * <h3>Implementation details</h3>
+   *
+   * In addition to the basic functionality provided by the QPManagerBase class,
+   * this class implements transferring of the quadrature point data between
+   * cells if case of adaptive refinement. That also includes automatic shipping of
+   * information between different processors.
+   *
+   * To that end, the constructor of the class is provided with three main objects:
+   * scalar FiniteElement @pprojection_fe, @p lhs_quadrature and @p rhs_quadrature Quadrature rules.
+   * First, the data located at @p rhs_quadrature is L2-projected to the FiniteElement.
+   * In  doing so the mass matrix of this element is required, which will be calculated
+   * with the @p lhs_quadrature rule . Then the data is sent to the possibly new owner of the
+   * cell. The class makes use of a feature of p4est (and parallel::distributed::Triangulation)
+   * that allows to attach information to cells during mesh refinement and rebalancing.
+   * On receiving information on the target cell, the data is projected back to the
+   * quadrature points. In case local refinement is performed, we have to first
+   * project local DoF values of the parent element to each child.
+   *
+   * Since this class is templated by DATA type, the users DATA class has to
+   * implement the following functions shown below for a quadrature point data
+   * with 2 scalars :
+   * <code>
+   * struct MyQData
+   * {
+   *   // a function which returns the total number of scalars which will be transfered
+   *   static unsigned int number_of_scalars()
```

Comment:
If we could find a way to make a run-time computable value (wherever its used) then that would be very helpful. This is important when the material constitution is not known _a priori_ (e.g. the number of internal variables being specified in a parameter file)